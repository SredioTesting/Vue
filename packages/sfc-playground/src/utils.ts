export function debounce(fn: Function, n = 100) {
  let handle: number | null = null
  return (...args: any[]) => {
    if (handle) window.clearTimeout(handle)
    handle = window.setTimeout(() => {
      fn(...args)
      handle = null
    }, n)
  }
}

// prefer old unicode hacks for backward compatibility
// https://base64.guru/developers/javascript/examples/unicode-strings
export function utoa(data: string): string {
  return btoa(unescape(encodeURIComponent(data)))
}

export function atou(base64: string): string {
  return decodeURIComponent(escape(atob(base64)))
}
