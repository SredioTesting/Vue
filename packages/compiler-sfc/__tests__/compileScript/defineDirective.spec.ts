import { assertCode, compileSFCScript as compile } from '../utils'

describe('defineProps', () => {
  test('basic usage', () => {
    const { content } = compile(`
    <script setup>
      const vTest = defineDirective({ created(){} })
    </script>
    `)
    expect(content).toMatchInlineSnapshot(`
      "
      export default {
        setup(__props, { expose: __expose }) {
        __expose();

            const vTest = { created(){} }
          
      return { vTest }
      }

      }"
    `)

    assertCode(content)
    expect(content).not.toMatch('defineDirective')
    expect(content).toMatch('const vTest = { created(){} }')
    expect(content).toMatch('return { vTest }')
  })

  test('with typescript', () => {
    const { content } = compile(`
    <script setup lang="ts">
      const vTest = defineDirective({ created(){} } as { name: string })
    </script>
    `)
    expect(content).toMatchInlineSnapshot(`
      "import { defineComponent as _defineComponent } from 'vue'

      export default /*#__PURE__*/_defineComponent({
        setup(__props, { expose: __expose }) {
        __expose();

            const vTest = { created(){} }
          
      return { vTest }
      }

      })"
    `)

    assertCode(content)
    expect(content).not.toMatch('defineDirective')
    expect(content).toMatch('const vTest = { created(){} }')
    expect(content).toMatch('return { vTest }')
  })

  test('use with other micro', () => {
    const { content } = compile(`
    <script setup>
      const vTest = defineDirective({ created(){} })
      const props = defineProps({
        bar: String
      })
    </script>
    `)
    expect(content).toMatchInlineSnapshot(`
      "
      export default {
        props: {
              bar: String
            },
        setup(__props, { expose: __expose }) {
        __expose();

            const vTest = { created(){} }
            const props = __props
          
      return { vTest, props }
      }

      }"
    `)

    assertCode(content)
    expect(content).not.toMatch('defineDirective')
    expect(content).toMatch('const vTest = { created(){} }')
    expect(content).toMatch('return { vTest, props }')
  })

  test('use of multiple define micro', () => {
    const { content } = compile(`
    <script setup>
      const vTest = defineDirective({ created(){} })
      const props = defineProps({
        bar: String
      })
      const vFoo = defineDirective({ mounted(){} })
    </script>
    `)
    expect(content).toMatchInlineSnapshot(`
      "
      export default {
        props: {
              bar: String
            },
        setup(__props, { expose: __expose }) {
        __expose();

            const vTest = { created(){} }
            const props = __props
            const vFoo = { mounted(){} }
          
      return { vTest, props, vFoo }
      }

      }"
    `)

    assertCode(content)
    expect(content).not.toMatch('defineDirective')
    expect(content).toMatch('const vTest = { created(){} }')
    expect(content).toMatch('const vFoo = { mounted(){} }')
    expect(content).toMatch('return { vTest, props, vFoo }')
  })
})
