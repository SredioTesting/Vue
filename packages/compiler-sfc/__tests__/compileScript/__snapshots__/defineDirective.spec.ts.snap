// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`defineProps > basic usage 2`] = `
"
export default {
  setup(__props, { expose: __expose }) {
  __expose();

      const vTest = { created(){} }
    
return { vTest }
}

}"
`;

exports[`defineProps > use of multiple define micro 2`] = `
"
export default {
  props: {
        bar: String
      },
  setup(__props, { expose: __expose }) {
  __expose();

      const vTest = { created(){} }
      const props = __props
      const vFoo = { mounted(){} }
    
return { vTest, props, vFoo }
}

}"
`;

exports[`defineProps > use with other micro 2`] = `
"
export default {
  props: {
        bar: String
      },
  setup(__props, { expose: __expose }) {
  __expose();

      const vTest = { created(){} }
      const props = __props
    
return { vTest, props }
}

}"
`;

exports[`defineProps > with typescript 2`] = `
"import { defineComponent as _defineComponent } from 'vue'

export default /*#__PURE__*/_defineComponent({
  setup(__props, { expose: __expose }) {
  __expose();

      const vTest = { created(){} }
    
return { vTest }
}

})"
`;
